### DISCUSSION TITLE
Getting information about Scheduled Jobs

### DISCUSSION THREAD

**Discussion by ejannett** [jobrunr/jobrunr#430] [Q&A] [answered]
Hi,

I am looking for a way to get back information about scheduled jobs. 
I use BackgroundJobRequest.scheduleRecurrently and BackgroundJobRequest.schedule APIs to trigger executions of things.  
I am looking for a programmatic way of getting information like what displayed in the dashboard within Scheduled jobs details page
: duration , last run time , next run time etc... 

The only APIs I see to fetch information about a given Job is the StorageProvider interface. Is this really public ? 
 First it seems not to cover all my needs 
and second I do not see how fill the id parameter of the getJobById APIs (not all the API to enqueue or scheduled jobs return Job object)

Any help would be greatly appreciated. 
thank in advance
manu

**Comment by rdehuyss** [comment, marked as answer]
Hi Manu,

your question is not 100% clear to me and if possible, I would like to know the business reasons behind your question: is this to make another type of dashboard, or is it to do predictions, ... ? Also, as far as I know, the Scheduled jobs details page does have last run time and next run time. This is in the Recurring Jobs page.

__I'll try my best to answer:__
All API's that create a `Job` like `enqueue` and `schedule` (except for when you create a lot of jobs using a Java 8 `Stream` and also not if you create a CRON `RecurringJob`) return a `JobId`. You can use that `JobId` to get the details of a Job.

For some special business cases you can indeed use the `StorageProvider` - it's one of the most tested things in JobRunr and the API is quite stable.

[JobRunr Pro](https://www.jobrunr.io/en/documentation/pro/jobrunr-pro-dashboard/) has a dashboard with more searching capabilities and the API of the `StorageProvider` is more extensive and allows to search like this:

```java
final Page<Job> jobs = storageProvider.getJobs(
    jobs()
        .withState(SUCCEEDED)
        .createdByRecurringJob("my-recurring-job-id")
        .build(), 
    ascOnUpdatedAt(10));
```

This will return the first 10 `Succeeded` jobs that were created by the recurring job with id `my-recurring-job-id` and ordered by updatedAt.  
I hope all of this makes sense?

**Comment by ejannett** [comment]
Hi,
First of all thank for the answer. 

In my application I have to trigger execution of tasks. This can be "immediate", postponed or recurring executions. 
For this I respectively use  _BackgroundJobRequest.enqueue(UUID id, JobRequest jobRequest)_, _BackgroundJobRequest.schedule(UUID id, Instant instant, JobRequest jobRequest)_ and _BackgroundJobRequest.scheduleRecurrently(String id, String cron, JobRequest jobRequest)_

I use JobRequest as what I need to scheduled cannot be express with simple lambdas. 

Another part of my application is about a simple management of what's been executed (let's agree on using this generic term.)
I am not building another dashboard, this is done as part of one of mine existing service.

This management is about : 

1. explicitly delete a job
2. getting the last execution time of a job
3. getting the next execution time of a job 
4. getting some outcome of a job

For #4 I guess that I would need JobResult but that part of JobRunr pro

For #1 I would  use BackgroundJobRequest.delete(JobID)

For #2 and #3 the only API I see is the StorageProvider interface using things like 
_StorageProvider.getJobById(<jobID>).getJobState()
StorageProvider.getJobById(<jobID>).getCreatedAt()
StorageProvider.getJobById(<jobID>).getUpdatedAt()_

but , first I do not see how this give the last execution time and the next execution time 
and then all these rely on knowing the **JobId**.
 _BackgroundJobRequest.enqueue(UUID id, JobRequest jobRequest)_ and _BackgroundJobRequest.schedule(UUID id, Instant instant, JobRequest jobRequest)_  give me back JobIDs 
but _BackgroundJobRequest.scheduleRecurrently(...)_  (the method I use the most) do not

The piece of code you've pasted above seems helpful but using JobRunrPRO is not an option for me 

thank in advance
manu
